imap/jmap_mailbox.c

```diff
@@ -2882,11 +2876,88 @@ static int _mbox_on_destroy_move(jmap_req_t *req,
         result->err = jmap_server_error(r);
     }
     jmap_closembox(req, &dst_mbox);
-    jmap_closembox(req, &src_mbox);
     ptrarray_fini(&move_msgrecs);
     return r;
 }
 
+static int find_bumpuids_cb(const conv_guidrec_t *rec, void *rock)
+{
+    struct seqset *bumpuids = rock;
+
+    if (rec->version < 1) {
+        syslog(LOG_ERR, "%s:%s: outdated guid record in mailbox %s",
+                __FILE__, __func__, rec->mboxname);
+        return IMAP_INTERNAL;
+    }
+    if (!(rec->system_flags & FLAG_DELETED) &&
+        !(rec->internal_flags & FLAG_INTERNAL_EXPUNGED)) {
+        seqset_add(bumpuids + rec->foldernum, rec->uid, 1);
+    }
+
+    return 0;
+}
+
+static int _mbox_destroy_bump_emailstate(jmap_req_t *req, struct mailbox *mbox)
+{
+    int numfolders = conversations_num_folders(req->cstate);
+    if (numfolders < 1) return 0;
+
+    struct seqset *bumpuids = xzmalloc(numfolders * sizeof(struct seqset));
+    int i;
+    for (i = 0; i < numfolders; i++) {
+        seqset_init(bumpuids + i, 0, SEQ_SPARSE);
+    }
+    struct mailbox_iter *iter = mailbox_iter_init(mbox, 0, ITER_SKIP_EXPUNGED);
+    struct buf guid = BUF_INITIALIZER;
+    const message_t *msg;
+    int r = 0;
+
+    /* For each message in the to-be destroyed mailbox, find the
+     * folder names and uids of all messages with the same guid. */
+    while ((msg = mailbox_iter_step(iter))) {
+        const struct index_record *record = msg_record(msg);
+        buf_setcstr(&guid, message_guid_encode(&record->guid));
+        int r = conversations_guid_foreach(req->cstate, buf_cstring(&guid),
+                                           find_bumpuids_cb, bumpuids);
+        if (r) break;
+    }
+    buf_free(&guid);
+    mailbox_iter_done(&iter);
+    if (r) goto done;
+
+    /* Bump messages by folder */
+    for (i = 0; i < numfolders; i++) {
+        struct seqset *uids = bumpuids + i;
+        if (seqset_first(uids)) {
+            struct mailbox *folder = NULL;
+            r = jmap_openmbox(req, conversations_folder_name(req->cstate, i),
+                              &folder, 1);
+            if (!r) {
+                uint32_t uid;
+                for (uid = seqset_first(uids); uid; uid = seqset_getnext(uids)) {
+                    struct index_record record;
+                    r = mailbox_find_index_record(folder, uid, &record);
+                    if (!r) {
+                        // XXX use same modseq for all bumps?
+                        r = mailbox_rewrite_index_record(folder, &record);
+                    }
+                    if (r) break;
+                }
+            }
+            jmap_closembox(req, &folder);
+            if (r) break;
+        }
+        if (r) break;
+    }
+
+done:
+    for (i = 0; i < numfolders; i++) {
+        seqset_fini(bumpuids + i);
+    }
+    free(bumpuids);
+    return r;
+}
+
```

Comment:
This is the relevant code section for review. The remaining changes are due to the renaming of seqset_init to seqset_new for naming consistency.